### PR TITLE
Scenario service: fix visibility of contract keys. 

### LINKED ISSUE #5924: Incorrect contract key lookups in Scenarios
On branch `contract-key-docs-WIP` I created an example to illustrate how contract keys behave for different relationships between Parties and Contracts. It highlighted a few issues:

1. Some of the `submitMustFail` statements fail with low-level crashes similar to that described in #5903 
2. There are two lookups that I consider wrong. https://github.com/digital-asset/daml/blob/contract-key-docs-WIP/docs/source/daml/code-snippets/Keys.daml#L133-L146. Non-stakeholders should always get negative lookup results at interpretation time, which means positive lookups can never succeed.

The script version of 2. (below) fails correctly meaning scenarios and Sandbox are mismatched. All of this was tested on the latest 1.1.0 snapshot.

```
s : Script () = do

  -- Put four parties in the four possible relationships with a `Keyed`
  sig <- allocateParty "s" -- Signatory
  obs <- allocateParty "o" -- Observer
  divulgee <- allocateParty "d" -- Divulgee
  blind <- allocateParty "b" -- Blind

  keyedCid <- submit sig do createCmd Keyed with ..
  divulgercid <- submit divulgee do createCmd Divulger with ..
  submit sig do exerciseCmd divulgercid DivulgeKeyed with ..

  -- Now the signatory and observer delegate their choices
  sigDelegationCid <- submit sig do 
    createCmd Delegation with 
      sig
      delegees = [obs, divulgee, blind]
  obsDelegationCid <- submit obs do 
    createCmd Delegation with
      sig = obs
      delegees = [divulgee, blind]

  -- TESTING LOOKUPS AND FETCHES

  -- WRONG: Divulgee can do positive lookup with maintainer authority. 
  -- The lookup should return `None`.
  mcid <- submit divulgee do 
    exerciseCmd sigDelegationCid LookupKeyed with
      delegee = divulgee
      lookupKey = sig
  assert (mcid == Some keyedCid)
  -- WRONG: Divulgee can fetch with stakeholder authority
  -- Should fail as lookup is negative.
  (cid, keyed) <- submit divulgee do 
    exerciseCmd obsDelegationCid FetchKeyed with
      delegee = divulgee
      lookupKey = sig
  assert (keyedCid == cid)

  return ()
```

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
fixes #5924

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 1]
```diff
@@ -217,7 +217,8 @@ final case class ScenarioRunner(
             missingWith(SErrorCrash(s"contract $acoid not effective, but we found its key!"))
           case LookupContractNotActive(_, _, _) =>
             missingWith(SErrorCrash(s"contract $acoid not active, but we found its key!"))
-          case LookupContractNotVisible(_, _, _) =>
+          case x @ (LookupOk(_, _, _) | LookupContractNotVisible(_, _, _)) =>
+            remy.log(x)
```
Comment (SamirTalwar-DA) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:221]: Whoops.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 2]
```diff
@@ -217,7 +217,7 @@ final case class ScenarioRunner(
             missingWith(SErrorCrash(s"contract $acoid not effective, but we found its key!"))
           case LookupContractNotActive(_, _, _) =>
             missingWith(SErrorCrash(s"contract $acoid not active, but we found its key!"))
-          case LookupContractNotVisible(_, _, _) =>
+          case LookupOk(_, _, _) | LookupContractNotVisible(_, _, _) =>
```
Comment (stefanobaghino-da) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:220]: It's probably worth mentioning why a `LookupOk` ends up here.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 3]
```diff
@@ -217,7 +217,7 @@ final case class ScenarioRunner(
...
             if (!cb(SKeyLookupResult.NotVisible))
               throw SErrorCrash(s"contract $acoid not visible, but we found its key!")
```
Comment (stefanobaghino-da) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:222]: Is this code valid to handle the new case?

File: daml-lf/tests/ContractKeys.daml [hunk 4]
```diff
@@ -0,0 +1,200 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
```
Comment (hurryabit) [inline daml-lf/tests/ContractKeys.daml:6]:
```suggestion
```

**Comment by remyhaemmerle-da** [comment]
@hurryabit wrote:
> I find it a bit worrying that the scenario service and the sandbox have separate implementations for this. We should aim for unifying the implementations in order to avoid diverging behavior.

Two remarks: 
* We should probably depose the scenario service and rely on something like daml-script + daml-on-memory for running the scenario in the IDE.

**Comment by bame-da** [comment]
> We should probably depose the scenario service and rely on something like daml-script + daml-on-memory for running the scenario in the IDE.

That's the dream.

**Comment by cocreature** [comment]
> We should probably depose the scenario service and rely on something like daml-script + daml-on-memory for running the scenario in the IDE.

That’s not feasible until we have an API for replacing individual modules in a package. With the current package API I have to reupload the whole package which is slow and I cannot garbage collect anything which is a memory leak. I still want DAML Script in the IDE but making it work without something like a scenario service is quite a bit of work.

**Pull request merged**